### PR TITLE
feat: allow for non-string columns in `select` for pandas backend

### LINKED ISSUE #827: Let non-string names "pass-through" in DataFrame.select
I think
```python
df = pd.DataFrame({0: [1, 2, 3], 1: [4,5,6]})
print(nw.from_native(df, eager_only=True).select(0).pipe(nw.to_native))
```
should be supported

(plus, something like this is used in sklearn)

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #827 

## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

### CODE REVIEW

File: docs/other/column_names.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+create one with _two_ columns named like that!
+
+What does Narwhals do about this?
+
+- In general, non-string column names are supported. In some places where this might
+  create ambiguity (such as `DataFrame.__getitem__`) we may be strict and only
+  allow passing in column names if they're strings. If you have a use-case that's
+  failing for non-string column names, please report it to [https://github.com/narwhals-dev/narwhals/issues](https://github.com/narwhals-dev/narwhals/issues)
+  and we'll see if we can support it.
+- Duplicate column names are 🚫 banned 🚫.
```
Comment (FBruzzesi) [inline docs/other/column_names.md:23]: I was so uncomfortable until here 😂

File: narwhals/_expression_parsing.py [hunk 2]
```diff
@@ -6,6 +6,7 @@
 from copy import copy
 from typing import TYPE_CHECKING
 from typing import Any
+from typing import Hashable
```
Comment (FBruzzesi) [inline narwhals/_expression_parsing.py:9]: Shall we adopt `collections.abc.Hashable`? [typing.Hashable](https://docs.python.org/3/library/typing.html#typing.Hashable) is getting deprecated

File: docs/other/column_names.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+Oh...not only does it let us create a dataframe a column named `0` - it lets us
+create one with _two_ columns named like that!
+
+What does Narwhals do about this?
+
+- In general, non-string column names are supported. In some places where this might
+  create ambiguity (such as `DataFrame.__getitem__`) we may be strict and only
+  allow passing in column names if they're strings. If you have a use-case that's
+  failing for non-string column names, please report it to [https://github.com/narwhals-dev/narwhals/issues](https://github.com/narwhals-dev/narwhals/issues)
+  and we'll see if we can support it.
```
Comment (FBruzzesi) [inline docs/other/column_names.md:22]: Fancy having this in its own bullet?

> If you have a use-case that's failing for non-string column names, please report it to [https://github.com/narwhals-dev/narwhals/issues](https://github.com/narwhals-dev/narwhals/issues) and we'll see if we can support it.

**Review by FBruzzesi — COMMENTED** [review]
I somehow thought we were already supporting this. I remember discussing such possibility before. Thanks for taking care of it🎉

Left a couple of considerations along the way but do not consider those as blocker in any way if you wanna merge it quickly

**Comment by MarcoGorelli** [comment]
thanks for your review, all good points!

in theory we'd decided to support this, but it looks like there were a few parts missing (which I noticed testing out scikit-learn)

File: docs/other/column_names.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+```python
+>>> import pandas as pd
+>>> pd.concat([pd.Series([1, 2], name=0), pd.Series([1, 3], name=0)], axis=1)
+   0  0
+0  1  1
+1  2  3
+```
+
+Oh...not only does it let us create a dataframe a column named `0` - it lets us
```
Comment (MarcoGorelli) [inline docs/other/column_names.md:13]:
```suggestion
Oh...not only does it let us create a dataframe with a column named `0` - it lets us
```

File: docs/other/column_names.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+create one with _two_ columns named like that!
```
Comment (MarcoGorelli) [inline docs/other/column_names.md:14]:
```suggestion
Oh...not only does it let us create a dataframe with a column named `0` - it lets us
create one with _two_ such columns!
```
